learning_observer/learning_observer/lms_integration.py

```diff
@@ -0,0 +1,287 @@
+    def arguments(self):
+        return extract_parameters_from_format_string(self.remote_url)
+
+    def _local_url(self):
+        parameters = "}/{".join(self.arguments())
+        base_url = f"/{self.lms}/{self.name}"
+        if len(parameters) == 0:
+            return base_url
+        else:
+            return base_url + "/{" + parameters + "}"
+
+    def _add_cleaner(self, name, cleaner):
+        if self.cleaners is None:
+            self.cleaners = dict()
+        self.cleaners[name] = cleaner
+        if 'local_url' not in cleaner:
+            cleaner['local_url'] = self._local_url + "/" + name
+
+    def _cleaners(self):
+        if self.cleaners is None:
+            return []
+        else:
+            return self.cleaners
+
+def extract_parameters_from_format_string(format_string):
+    '''
+    Extracts parameters from a format string. E.g.
+    >>> ("hello {hi} my {bye}")]
+    ['hi', 'bye']
+    '''
+    return [f[1] for f in string.Formatter().parse(format_string) if f[1] is not None]
+
+def raw_access_partial(raw_ajax_function, target_url, name=None):
+    '''
+    This is a helper which allows us to create a function which calls specific
+    Google APIs.
+
+    To test this, try:
+
+        print(await raw_document(request, documentId="some_google_doc_id"))
+    '''
+    async def caller(request, **kwargs):
+        '''
+        Make an AJAX request to LMS
+        '''
+        return await raw_ajax_function(request, target_url, **kwargs)
+    setattr(caller, "__qualname__", name)
+
+    return caller
+
+def api_docs_handler(endpoints):
+    '''
+    Return a list of available endpoints.
+
+    Eventually, we should also document available function calls
+    '''
+    response = "URL Endpoints:\n\n"
+    for endpoint in endpoints:
+        response += f"{endpoint._local_url()}\n"
+        cleaners = endpoint._cleaners()
+        for c in cleaners:
+            response += f"   {cleaners[c]['local_url']}\n"
+    response += "\n\n Globals:"
+    return aiohttp.web.Response(text=response)
+
+def register_cleaner(data_source, cleaner_name, endpoints):
+    '''
+    This will register a cleaner function, for export both as a web service
+    and as a local function call.
+    '''
+    def decorator(f):
+        found = False
+        for endpoint in endpoints:
+            if endpoint.name == data_source:
+                found = True
+                endpoint._add_cleaner(
+                    cleaner_name,
+                    {
+                        'function': f,
+                        'local_url': f'{endpoint._local_url()}/{cleaner_name}',
+                        'name': cleaner_name
+                    }
+                )
+
+        if not found:
+            raise AttributeError(f"Data source {data_source} invalid; not found in endpoints.")
+        return f
+
+    return decorator
+
+def make_ajax_raw_handler(raw_ajax_function, remote_url):
```

Comment:
At the very least, the docstring should explain enough to be able to use this _securely_.